wallet/src/key.rs

```diff
@@ -78,22 +80,130 @@ where
         let (sk_bytes, chain_code_bytes) = result.split_at(32);
 
         // secret/chain_code computation might panic if length returned by hmac is wrong
-        let secret = secp256k1::SecretKey::from_slice(sk_bytes).expect("Secret Key length error");
+        let secret_key = SecretKey::from_slice(sk_bytes).expect("Secret Key length error");
         let mut chain_code = [0u8; 32];
         chain_code.copy_from_slice(chain_code_bytes);
 
-        Ok(ExtendedKey { secret, chain_code })
+        Ok(ExtendedSK {
+            secret_key,
+            chain_code,
+        })
     }
 }
 
+#[derive(Debug, PartialEq, Fail)]
+pub enum KeyDerivationError {
+    /// Invalid hmac key length
+    #[fail(display = "The length of the hmac key is invalid")]
+    InvalidKeyLength,
+    /// Invalid seed length
+    #[fail(display = "The length of the seed is invalid, must be between 128/256 bits")]
+    InvalidSeedLength,
+
+    #[fail(display = "Error in secp256k1 crate")]
+    Secp256k1Error,
+}
+
 /// Extended Key is just a Key with a Chain Code
-pub struct ExtendedKey<K> {
-    pub secret: K,
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ExtendedSK {
+    pub secret_key: SecretKey,
     pub chain_code: [u8; 32],
 }
 
+/// A child number for a derived key
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct ChildNumber(u32);
+
+impl ChildNumber {
+    pub fn is_hardened(&self) -> bool {
+        self.0 & HARDENED_BIT == HARDENED_BIT
```

Comment:
Assuming that a ChildNumber cannot be normal and hardened at the same time:

```suggestion
        !self.is_normal()
```